### DISCUSSION TITLE
Does/Could Ibis ingest substrait plans?

### DISCUSSION THREAD

**Discussion by jonmmease** [ibis-project/ibis-substrait#840] [Ideas]
Apologies if this is discussed elsewhere (or already possible and I just missed it), but I'm wondering if it would be possible to input a substrait plan to build an Ibis expression.  The use case I have in mind is wondering if there's a future where [VegaFusion](https://github.com/hex-inc/vegafusion) could generate substrait plans internally, build Ibis expressions from them, and then let Ibis handle the SQL generation for systems that don't support substrait.

I've been chatting with @jcrist about integrating Ibis with VegaFusion using VegaFusion's existing extension points in https://github.com/hex-inc/vegafusion/discussions/355, but after learning more about substrait I'm wondering if this would be an alternative future integration path.

**Comment by gforsyth** [comment]
Hey @jonmmease -- we used to have a "decompiler" to go from substrait expressions to ibis expressions -- initially we built it because there were no substrait consumers so we had to be both a producer and a consumer to test anything.
But after there were consumers, there wasn't much of a reason for us to keep it around.
If it would be useful, it's something we can try to resurrect, but I'm not sure when exactly we might be able to make that happen.

Here's the PR where I ripped out the decompiler (which was already out-of-date and broken at time of removal): https://github.com/ibis-project/ibis-substrait/pull/692

**Reply by jonmmease** [reply]
Thanks for the helpful context @gforsyth!  That all makes sense. I might poke around to see how things used to work.

In general, I think it would be nice for substrait adoption if there were a way to convert substrait plans to SQL for backends that don't support it yet. This would make substrait a really appealing target for higher-level systems that would otherwise need to deal with SQL dialects themselves.  Does it make sense for Ibis to eventually play this role? Or are there reasons it's not a good fit?

**Reply by gforsyth** [reply]
I'm not sure, to be honest.  Ibis is definitely one possible path -- but it's also its own abstraction layer and we don't always generate the cleanest SQL (although it's getting better).

I can definitely see a role for a direct Substrait -> SQL transpiler, although then you'd still be stuck with the SQL dialect problem.

**Reply by jonmmease** [reply]
I don't see clean SQL as a priority, or even SQL in and of itself. Just having an adaptor between substrait and the "before substrait" ecosystem of compute engines would be really useful.

**Comment by tokoko** [comment]
@gforsyth I'm also working on a feature in [feast](https://feast.dev/) where I need to store user-provided functions in substrait and be able to run them on 1) non-python substrait consumers directly and also 2) bring them back to ibis to extend already existing ibis transformations. (described [here](https://github.com/feast-dev/feast/issues/3945)). Are you against bringing back the decompiler for some reason or is the maintenance burden the only issue?

P.S. You mentioned "Substrait -> SQL transpiler" above... I remember reading on an ibis page that the plan was to drop internal representation and use substrait instead, which will make ibis itself substrait -> sql transpiler in a way. Is that no longer the plan?

**Reply by gforsyth** [reply]
Hey @tokoko -- it's the maintenance burden.  If we do bring it back, it will be after we release Ibis 9.0 -- there was a significant IR rewrite in Ibis (current `main`) and rebuilding the compiler now would be mostly wasted work.  We _think_ that the Ibis IR is stabilizing now, and if that proves to be true, then that would reduce the maintenance burden of keeping ibis-substrait up-to-date.

I don't recall any plans to completely drop the Ibis IR -- currently I don't think that Substrait can necessarily represent everything we try to track, but also we would need to write yet another transpiler to shove Substrait IR into the sqlglot IR to get to sql generation.

I think the more likely long-term result is using Ibis-substrait to take the Ibis IR, generate a substrait plan, and then pass that directly to a backend (that obviously needs to support substrait) and skip the SQL generation altogether.

**Reply by tokoko** [reply]
hey, thanks for the response. I think that was my wrong interpretation of this sentence "In the long-term, we aim for a standard query plan Intermediate Representation (IR) like [Substrait](https://substrait.io/) to simplify this further." on ibis [why](https://ibis-project.org/why) page.

Don't want to be too much bother to you, but you're probably the best person to ask this :). Can you point me to some general areas where substrait and ibis IRs fundamentally diverge from one another? I'm trying to understand if/when both compiler and decompiler are in `ibis-substrait`, what are the types of queries for which the process of back-to-back compilation and decompilation won't work (correctly).

and a quick last question.. substrait was also mentioned in this [discussion](https://github.com/ibis-project/ibis/discussions/7704) (by me and couple others) as a way to serialize ibis IR. Is that realistic? Or does a stable ibis IR mean that it will probably at some point have it's own serialized representation?

**Reply by gforsyth** [reply]
> I think that was my wrong interpretation of this sentence "In the long-term, we aim for a standard query plan Intermediate Representation (IR) like [Substrait](https://substrait.io/) to simplify this further." on ibis [why](https://ibis-project.org/why) page.

Ahh, gotcha.  Yeah, that sentence can probably be rephrased to be clearer.  Thanks for pointing that out!

>  Can you point me to some general areas where substrait and ibis IRs fundamentally diverge from one another?

They are largely compatible in terms of the general building blocks.  I don't think Substrait has formalized its thoughts on semi- and anti-joins yet, although that isn't the most common pattern anyway.  
I don't think there are any queries where we _couldn't_ round-trip the queries, there's just some difficult bookkeeping to do along the way.  Most of that is around how substrait handles subqueries and also the way that it flattens all colums of all tables into a single list.  I don't think I ever got proper subquery reconstruction working, just because tracking which columns were associated with which subquery was (at least at the time) very difficult.

> and a quick last question.. substrait was also mentioned in this https://github.com/ibis-project/ibis/discussions/7704 (by me and couple others) as a way to serialize ibis IR. Is that realistic? Or does a stable ibis IR mean that it will probably at some point have it's own serialized representation?

We're firm believers in open standards, and if there's a standard that can represent an ibis expression with good fidelity, we'll use that before rolling our own.  I would be very surprised if we ended up with an ibis-specific serialized representation.

**Reply by jonmmease** [reply]
> ... but also we would need to write yet another transpiler to shove Substrait IR into the sqlglot IR to get to sql generation.

Huh, I hadn't thought of this approach. I wonder if this would make more sense as a standalone project for substrait -> sql conversion.  @gforsyth, from your experience working with sqlglot and substrait, does this seem like a tractable approach in terms of the level of abstraction between the two?

This might be in interesting approach

**Reply by gforsyth** [reply]
Hey @jonmmease!  Yeah, if I were going to write a Substrait -> SQL I would definitely reach for sqlglot. 
I think the abstraction levels are comparable.  I _would_ recommend figuring out a topological sort for substrait plans (specifically with regards to subqueries) because things get gnarly when recursing and trying to manage column references.

**Reply by tokoko** [reply]
This might prove fruitless, but I recently started a side-project [here](https://github.com/tokoko/subframe) trying to implement ibis API without either ibis IR or sqlglot IR, doing everything directly on substrait. While most of these stuff is already possible with a set of compilers and decompilers, it feels (at least to me) a waste having three different IRs with very similar goals (abstract backend-less IR) instead of just using one. I'm curious to see how far I can go with this.
